### ISSUE TITLE
Bug: no-unused-expressions incorrectly triggers when using comma operator

### ISSUE THREAD

**Issue by STRd6** [eslint/eslint#18758]
### Environment

Local ESLint version: v9.8.0

### What parser are you using?

Default (Espree)

### What did you do?

```js
/*eslint no-unused-expressions: "error"*/
f(), g()
```

### What did you expect to happen?

No error because all expressions are used or have effects

### What actually happened?

Error: Expected an assignment or function call and instead saw an expression.  ([no-unused-expressions](https://eslint.org/docs/rules/no-unused-expressions))

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLWV4cHJlc3Npb25zOiBcImVycm9yXCIqL1xuXG5mKCksIGcoKSIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsic291cmNlVHlwZSI6Im1vZHVsZSIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

This looks like it was working in v8.57.0 but got broken somewhere between there and v9.8.0

There's already a rule to disallow the comma operator https://eslint.org/docs/latest/rules/no-sequences

The rule for no-unused-expressions should recurse into the comma operator and only trigger if any sub-expression is unused

```
1, f() // should trigger the error, 1 is an unused expression
f(), g() // should not trigger
a = 1, b() // should not trigger
```

**Comment by aladdin-add** [comment]
Sounds like a bug to me.

**Comment by mdjermanovic** [comment]
> ```js
> /*eslint no-unused-expressions: "error"*/
> f(), g()
> ```

> Error: Expected an assignment or function call and instead saw an expression.

Per the [docs](https://eslint.org/docs/latest/rules/no-unused-expressions), this behavior is intentional:

https://github.com/eslint/eslint/blob/aac191e6701495666c264f71fc440207ea19251f/docs/src/rules/no-unused-expressions.md?plain=1#L29

> This looks like it was working in v8.57.0 but got broken somewhere between there and v9.8.0

I checked with v8.57.0 and the behavior is the same as with v9.8.0.

**Comment by STRd6** [comment]
Docs seem to indicate it is a bug:

`unless their return value is assigned or used in a condition evaluation, or a function call is made with the sequence expression value. `

**Comment by STRd6** [comment]
Some more info about the version. It may be the case that this always the behavior and the rule got added to @typescript-eslint/eslint-plugin between 7.8.0 and 8.0.1 revealing the issue.

https://github.com/bbrk24/Trilangle/pull/142/files#diff-2309320c1ba42d357abbeef63e4d65faf0be4131af576cd9ed198a36b27f6d1bL8-R10

**Comment by bbrk24** [comment]
I was able to reproduce this with typescript-eslint `^8.0.1` and eslint `~9.0.0`, so this isn't exactly a new thing from eslint's perspective.

**Comment by mdjermanovic** [comment]
> Docs seem to indicate it is a bug:
> 
> `unless their return value is assigned or used in a condition evaluation, or a function call is made with the sequence expression value. `

This refers to code such as `x = (f, g)` or `(f, g)()`, not `f(), g()`.



> Here's an example where the sequence is assigned and still gives an error: https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLWV4cHJlc3Npb25zOiBcImVycm9yXCIqL1xuXG54ID0gZigpLCBnKCkiLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InNvdXJjZVR5cGUiOiJtb2R1bGUiLCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19

`x = f(), g()` is the same as `(x = f()), g()`, so the value of the sequence expression is not assigned.

**Comment by molvqingtai** [comment]
遇到同样的报错，像这样：

if(isXXX){
  func()
}

**Comment by mdjermanovic** [comment]
> I encountered the same error, like this:
> 
> **work**
> ```
> if(isXXX){
>   func()
> }
> ``` 
>
> **error**
> 
> ```
> isXXX && func()
> ```

For this pattern, you can the `allowShortCircuit` option: https://eslint.org/docs/latest/rules/no-unused-expressions#allowshortcircuit

[Playground Demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLWV4cHJlc3Npb25zOiBbXCJlcnJvclwiLCB7IFwiYWxsb3dTaG9ydENpcmN1aXRcIjogdHJ1ZSB9XSovXG5cbmlzWFhYICYmIGZ1bmMoKSAvLyBubyBlcnJvciIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsic291cmNlVHlwZSI6Im1vZHVsZSIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=)

**Comment by mdjermanovic** [comment]
Closing as the current behavior is intentional and documented.
